File: segfault-examples/03_buffer_overflow.cpp
#include <iostream>
#include <vector>
#include <string.h>

// Example 1: Array out-of-bounds write
void arrayOverflow() {
    std::cout << "\n=== Array Buffer Overflow ===" << std::endl;
    
    int arr[5] = {1, 2, 3, 4, 5};
    std::cout << "Array of 5 elements" << std::endl;
    
    // CRASH: Writing way past array bounds
    std::cout << "Writing to arr[100]..." << std::endl;
    arr[100] = 42;
    
    std::cout << "This may or may not crash immediately" << std::endl;
}

// Example 2: Vector unchecked access
void vectorUncheckedAccess() {
    std::cout << "\n=== Vector Unchecked Access ===" << std::endl;
    
    std::vector<int> vec = {1, 2, 3};
    std::cout << "Vector with 3 elements" << std::endl;
    
    // CRASH: operator[] doesn't bounds-check!
    std::cout << "Accessing vec[10]..." << std::endl;
    vec[10] = 42;
    
    std::cout << "Undefined behavior" << std::endl;
}

// Example 3: String buffer overflow
void stringBufferOverflow() {
    std::cout << "\n=== String Buffer Overflow ===" << std::endl;
    
    char buffer[10];
    const char* longString = "This is a very long string that will overflow";
    
    std::cout << "Copying long string to small buffer..." << std::endl;
    
    // CRASH: strcpy doesn't check bounds!
    strcpy(buffer, longString);
    
    std::cout << "Buffer: " << buffer << std::endl;
}

// Example 4: Off-by-one error
void offByOne() {
    std::cout << "\n=== Off-By-One Error ===" << std::endl;
    
    int arr[10];
    std::cout << "Filling array of 10 elements..." << std::endl;
    
    // CRASH: Loop goes one past the end
    for (int i = 0; i <= 10; i++) {  // Should be i < 10
        arr[i] = i;
    }
    
    std::cout << "Off-by-one can cause corruption or crash" << std::endl;
}

// Example 5: Negative index
void negativeIndex() {
    std::cout << "\n=== Negative Index ===" << std::endl;
    
    int arr[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    std::cout << "Accessing arr[-5]..." << std::endl;
    
    // CRASH: Negative index accesses memory before array
    std::cout << "Value: " << arr[-5] << std::endl;
}

// Example 6: Stack buffer overflow
void stackBufferOverflow() {
    std::cout << "\n=== Stack Buffer Overflow ===" << std::endl;
    
    char buffer[8];
    const char* input = "VeryLongInputString";
    
    std::cout << "Using unsafe strcpy..." << std::endl;
    
    // CRASH: Classic stack buffer overflow
    strcpy(buffer, input);
}

// SAFE ALTERNATIVES
void safeAlternatives() {
    std::cout << "\n=== SAFE ALTERNATIVES ===" << std::endl;
    
    // 1. Vector with at() for bounds checking
    {
        std::cout << "\n1. Vector with at():" << std::endl;
        std::vector<int> vec = {1, 2, 3};
        
        try {
            vec.at(10) = 42;  // Throws exception instead of crashing
        } catch (const std::out_of_range& e) {
            std::cout << "  Caught exception: " << e.what() << std::endl;
        }
    }
    
    // 2. Use std::string instead of char arrays
    {
        std::cout << "\n2. Using std::string:" << std::endl;
        std::string str = "Hello";
        str += " World!";  // Safe concatenation
        std::cout << "  String: " << str << std::endl;
    }
    
    // 3. Use std::array with bounds checking
    {
        std::cout << "\n3. Using std::array:" << std::endl;
        std::array<int, 5> arr = {1, 2, 3, 4, 5};
        
        try {
            arr.at(10) = 42;  // Bounds checked
        } catch (const std::out_of_range& e) {
            std::cout << "  Caught exception: " << e.what() << std::endl;
        }
    }
    
    // 4. Safe string copy
    {
        std::cout << "\n4. Safe string copy:" << std::endl;
        char buffer[10];
        const char* source = "VeryLongString";
        
        strncpy(buffer, source, sizeof(buffer) - 1);
        buffer[sizeof(buffer) - 1] = '\0';
        
        std::cout << "  Buffer (truncated): " << buffer << std::endl;
    }
    
    // 5. Range-based loops
    {
        std::cout << "\n5. Range-based loops (can't overflow):" << std::endl;
        std::vector<int> vec = {1, 2, 3, 4, 5};
        
        for (int val : vec) {
            std::cout << "  " << val;
        }
        std::cout << std::endl;
    }
    
    // 6. Iterators with bounds checking
    {
        std::cout << "\n6. Checked iteration:" << std::endl;
        std::vector<int> vec = {1, 2, 3};
        
        for (size_t i = 0; i < vec.size(); i++) {
            std::cout << "  vec[" << i << "] = " << vec[i] << std::endl;
        }
    }
}

int main() {
    std::cout << "=== Buffer Overflow Segfault Examples ===" << std::endl;
    std::cout << "Uncomment ONE example to see the crash\n" << std::endl;
    
    // UNCOMMENT ONE TO SEE CRASH:
    // arrayOverflow();
    // vectorUncheckedAccess();
    // stringBufferOverflow();
    // offByOne();
    // negativeIndex();
    // stackBufferOverflow();
    
    // This is safe to run:
    safeAlternatives();
    
    std::cout << "\nProgram completed safely" << std::endl;
    std::cout << "Tip: Compile with -fsanitize=address to catch these bugs!" << std::endl;
    
    return 0;
}

File: segfault-examples/03_buffer_overflow.rs
// Rust prevents buffer overflows through bounds checking!

fn main() {
    println!("=== Buffer Overflow Safety in Rust ===\n");
    
    // Example 1: Array access is bounds-checked
    println!("Example 1: Array bounds checking");
    let arr = [1, 2, 3, 4, 5];
    println!("  Array: {:?}", arr);
    
    // Accessing within bounds
    println!("  arr[2] = {}", arr[2]);
    
    // This would PANIC at runtime (not compile error, but safe crash):
    // println!("  arr[100] = {}", arr[100]);  // PANIC: index out of bounds
    println!("  Out-of-bounds access would panic (safe crash, not undefined behavior)");
    
    // Example 2: Vec access is also bounds-checked
    println!("\nExample 2: Vec bounds checking");
    let vec = vec![1, 2, 3];
    println!("  Vec: {:?}", vec);
    
    // Safe access
    println!("  vec[1] = {}", vec[1]);
    
    // This would PANIC:
    // println!("  vec[10] = {}", vec[10]);  // PANIC: index out of bounds
    println!("  Out-of-bounds access causes panic, not memory corruption");
    
    // Example 3: Safe checked access with get()
    println!("\nExample 3: Safe access with get()");
    let vec = vec![1, 2, 3];
    
    match vec.get(10) {
        Some(val) => println!("  Value: {}", val),
        None => println!("  Index 10 is out of bounds (returned None)"),
    }
    
    // Example 4: Iterators prevent overflow
    println!("\nExample 4: Iterator safety");
    let arr = [1, 2, 3, 4, 5];
    
    for (i, val) in arr.iter().enumerate() {
        println!("  arr[{}] = {}", i, val);
    }
    println!("  Iterators can't go out of bounds!");
    
    // Example 5: Slices are bounds-checked
    println!("\nExample 5: Slice bounds checking");
    let arr = [1, 2, 3, 4, 5];
    let slice = &arr[1..4];
    println!("  Slice: {:?}", slice);
    
    // This would PANIC:
    // let bad_slice = &arr[1..100];  // PANIC: range end out of bounds
    println!("  Invalid slice ranges cause panic");
    
    // Example 6: String indexing is safe
    println!("\nExample 6: String safety");
    let s = String::from("Hello");
    
    // Can't directly index strings (prevents invalid UTF-8)
    // let c = s[0];  // COMPILE ERROR: cannot index into String
    
    // Must use safe methods
    if let Some(c) = s.chars().nth(0) {
        println!("  First char: {}", c);
    }
    
    // Example 7: No off-by-one errors with ranges
    println!("\nExample 7: Safe loops with ranges");
    let arr = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
    
    for i in 0..arr.len() {  // Range is exclusive of end
        print!("{} ", arr[i]);
    }
    println!("\n  Range syntax prevents off-by-one errors");
    
    // Example 8: Checked arithmetic
    println!("\nExample 8: Checked indexing");
    let vec = vec![1, 2, 3, 4, 5];
    let index: usize = 2;
    
    // Safe access with bounds check
    if index < vec.len() {
        println!("  vec[{}] = {}", index, vec[index]);
    }
    
    // Using get for Option return
    if let Some(val) = vec.get(index) {
        println!("  Using get: {}", val);
    }
    
    // Example 9: Array initialization is safe
    println!("\nExample 9: Safe array initialization");
    let arr = [0; 10];  // Array of 10 zeros
    println!("  Array length: {}", arr.len());
    println!("  All elements initialized safely");
    
    // Example 10: Buffer copying is safe
    println!("\nExample 10: Safe buffer operations");
    let mut dest = vec![0u8; 5];
    let src = vec![1u8, 2, 3, 4, 5, 6, 7, 8];
    
    // copy_from_slice checks lengths!
    if dest.len() <= src.len() {
        let len = dest.len(); dest.copy_from_slice(&src[0..len]);
        println!("  Safely copied: {:?}", dest);
    }
    
    // This would PANIC if lengths don't match:
    // dest.copy_from_slice(&src);  // PANIC: source slice length (8) != destination (5)
    
    println!("\n=== Key Takeaways ===");
    println!("1. All array/vec accesses are bounds-checked");
    println!("2. Out-of-bounds access panics (safe crash) not undefined behavior");
    println!("3. get() method returns Option for safe access");
    println!("4. Iterators prevent index errors");
    println!("5. Slice ranges are validated");
    println!("6. String indexing prevents invalid UTF-8");
    println!("7. Range syntax prevents off-by-one errors");
    println!("8. No silent memory corruption");
    println!("\nResult: Buffer overflows are prevented at RUNTIME!");
    println!("(And many are caught at compile time)");
}
